Modules/Scripted/SegmentStatistics/SegmentStatistics.py

```diff
@@ -422,20 +422,38 @@ def computeStatistics(self):
     self.reset()
 
     segmentationNode = slicer.mrmlScene.GetNodeByID(self.getParameterNode().GetParameter("Segmentation"))
-
-    # Get segment ID list
-    visibleSegmentIds = vtk.vtkStringArray()
-    if self.getParameterNode().GetParameter('visibleSegmentsOnly')=='True':
-      segmentationNode.GetDisplayNode().GetVisibleSegmentIDs(visibleSegmentIds)
-    else:
-      segmentationNode.GetSegmentation().GetSegmentIDs(visibleSegmentIds)
-    if visibleSegmentIds.GetNumberOfValues() == 0:
-      logging.debug("computeStatistics will not return any results: there are no visible segments")
-
-    # update statistics for all segment IDs
-    for segmentIndex in range(visibleSegmentIds.GetNumberOfValues()):
-      segmentID = visibleSegmentIds.GetValue(segmentIndex)
-      self.updateStatisticsForSegment(segmentID)
+    transformedSegmentationNode = None
+    try:
+      if not segmentationNode.GetParentTransformNode() is None:
+        # Create a temporary segmentation and harden the transform to ensure that the statistics are calculated
+        # in world coordinates
+        transformedSegmentationNode = slicer.vtkMRMLSegmentationNode()
+        transformedSegmentationNode.Copy(segmentationNode)
+        transformedSegmentationNode.HideFromEditorsOn()
+        slicer.mrmlScene.AddNode(transformedSegmentationNode)
+        transformedSegmentationNode.HardenTransform()
+        self.getParameterNode().SetParameter("Segmentation", transformedSegmentationNode.GetID())
+
+      # Get segment ID list
+      visibleSegmentIds = vtk.vtkStringArray()
+      if self.getParameterNode().GetParameter('visibleSegmentsOnly')=='True':
+        segmentationNode.GetDisplayNode().GetVisibleSegmentIDs(visibleSegmentIds)
+      else:
+        segmentationNode.GetSegmentation().GetSegmentIDs(visibleSegmentIds)
+      if visibleSegmentIds.GetNumberOfValues() == 0:
+        logging.debug("computeStatistics will not return any results: there are no visible segments")
+
+      # update statistics for all segment IDs
+      for segmentIndex in range(visibleSegmentIds.GetNumberOfValues()):
+        segmentID = visibleSegmentIds.GetValue(segmentIndex)
+        self.updateStatisticsForSegment(segmentID)
+    except:
+      logging.error("computeStatistics: Error when calculating statistics")
```

Comment:
This is good in the sense that now the parameter node remains consistent even if an error occurs. The only not nice thing is that now we swallow the error: we just log the error, no stack trace, and the caller method and the user has no way of knowing that something is wrong (other than no output is generated).

It would be nicer not to catch the exception or re-raise it (the simplest I think is to remove the "except:" branch and just keep "try" and "finally"). Try/except should be also added in the GUI to catch the error, display it to the user, and leave the module in a consistent state (similarly how it is done in current module template https://github.com/Slicer/Slicer/blob/master/Utilities/Templates/Modules/Scripted/TemplateKey.py#L270-L285). Now it does not work well - for example if an exception is thrown then the apply button gets stuck.

Sorry for just keep asking more but this part of the module was just not implemented well and it would be nice to fix it now that we touch it anyway. Thank you!

Reply:
Ok, thanks! Added the try/except block to the onApply function.